hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/InstantStateHandler.java

```diff
@@ -0,0 +1,131 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(InstantStateHandler.class);
+
+  /**
+   * Base url for  instant state requests.
+   */
+  private static final String BASE_URL = "/v1/hoodie/instantstate";
+
+  /**
+   * Param for  instant state requests, which contains a uniqueId for different writers.
+   */
+  public static final String INSTANT_STATE_DIR_PATH_PARAM = "instantstatedirpath";
+
+  /**
+   * GET requests. Returns all the instant states under instant state path.
+   */
+  public static final String ALL_INSTANT_STATE_URL = String.format("%s/%s", BASE_URL, "all");
+
+  /**
+   * POST requests. Refresh the instant state data cached in memory.
+   */
+  public static final String REFRESH_INSTANT_STATE = String.format("%s/%s", BASE_URL, "refresh/");
+
+  /**
+   * Cached instant state data, instant state path -> list of instant states in fs.
+   */
+  private final ConcurrentHashMap<String, List<InstantStateDTO>> cachedInstantStates;
+
+  /**
+   * Number of requests after the last refresh.
+   */
+  private final AtomicLong requestCount;
+
+  public InstantStateHandler(Configuration conf, TimelineService.Config timelineServiceConfig, FileSystem fileSystem,
+                             FileSystemViewManager viewManager) throws IOException {
+    super(conf, timelineServiceConfig, fileSystem, viewManager);
+    this.cachedInstantStates = new ConcurrentHashMap<>();
+    this.requestCount = new AtomicLong();
+  }
+
+  /**
+   * Read instant states from cache of file system.
+   *
+   * @return Instant states under the input instant state path.
+   */
+  public List<InstantStateDTO> getAllInstantStates(String instantStatePath) {
+    if (requestCount.incrementAndGet() >= timelineServiceConfig.instantStateForceRefreshRequestNumber) {
+      // Do refresh for every N requests to ensure the writers won't be blocked forever
+      refresh(instantStatePath);
+    }
+    return cachedInstantStates.computeIfAbsent(instantStatePath, k -> scanInstantState(new Path(k)));
+  }
+
+  /**
+   * Refresh the checkpoint messages cached. Will be called when coordinator start/commit/abort instant.
+   *
+   * @return Whether refreshing is successful.
+   */
+  public boolean refresh(String instantStatePath) {
+    try {
+      cachedInstantStates.put(instantStatePath, scanInstantState(new Path(instantStatePath)));
+      requestCount.set(0);
+    } catch (Exception e) {
+      LOG.error("Failed to load instant states, path: " + instantStatePath, e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Scan the instant states from file system.
+   */
+  public List<InstantStateDTO> scanInstantState(Path instantStatePath) {
+    try {
+      // Check instantStatePath exists before list status, see HUDI-5915
+      if (this.fileSystem.exists(instantStatePath)) {
+        return Arrays.stream(this.fileSystem.listStatus(instantStatePath)).map(InstantStateDTO::fromFileStatus).collect(Collectors.toList());
```

Comment:
So the server based ckp metadata still backed by the fs ckp message files ? Can you eliminate that part, just accepts the request from coordinator the store the instant metadata in memory should be enought ? There is no need to write into the fs anymore I think. Or do you think the http server is not stable so it will ack with bad response frequently?

Reply:
You are right, we can implement a complete timeline server memory-based checkpoint metadata since we will delete all previous ckp metadata when bootstrapping(so we do not actually rely on a durable filesystem state). And I think the timeline server is actually stable enough to serve the checkpoint message requests. 

For now, I think it's OK to keep the filesystem-backed checkpoint messages and make the timeline server a stateless reader for the following reasons: 
1. We will have a safe fallback option when the timeline server is not available during runtime. 
2. The filesystem-backed checkpoint code is still useful since we need to turn off the timeline server sometimes. And creating ckp messages files is not expensive, we can reuse the code. 
3. There will be no conflict if we want to move the start/abort/commit request to the timeline server in the future, and this PR has already implemented the important optimization that reduces the massive file listing. Moving the write part to the timeline server needs extra code and some more careful tests, we can move this PR. 

We can merge this PR and create a Jira to track any possible optimization, what do you think? And thanks again for your review~

Reply:
Please fire a JIRA to track this.